tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+      [&](const HloInstruction& insn) {
+        return absl::c_all_of(insn.users(), [&](const HloInstruction* user) {
+          if (user->opcode() == HloOpcode::kGetTupleElement) {
+            // Skip GTE.
+            return is_consumer_the_only_user(*user);
+          } else if (user == consumer) {
+            // user is consumer.
+            return true;
+          } else if (user == user->parent()->root_instruction()) {
+            // Consumed by ROOT is always fine, since it is impossible to create
+            // cycles through ROOT.
+            return true;
+          } else {
+            return false;
+          }
+        });
+      };
+
+  auto is_profitable = [](const HloInstruction& insn) -> bool {
+    // Do not fuse computation with too large shapes or too many instructions,
+    // because it may not be easily profitable.
+    constexpr int64 kShapeThreshold = 128 * 2048;
+    constexpr int64 kInsnCountThreshold = 30;
+    auto root = insn.fused_expression_root();
+
+    // Too large shapes are not easily profitable.
+    if (root->opcode() == HloOpcode::kTuple) {
+      // Since all output shapes are the same, use the first shape as the
+      // representative.
+      auto shape = root->operand(0)->shape();
+      if (ShapeUtil::ElementsIn(shape) > kShapeThreshold) {
+        return false;
+      }
+    } else {
+      auto shape = root->shape();
+      if (ShapeUtil::ElementsIn(shape) > kShapeThreshold) {
+        return false;
+      }
+    }
+
+    // Having too many instructions is not easily profitable.
+    if (insn.fused_instruction_count() > kInsnCountThreshold) {
+      return false;
+    }
+
+    return true;
+  };
+
+  // Do not fuse computations with non-row-major layouts because it may
+  // increase bandwidth pressure.
+  auto all_row_major = [](const HloInstruction& fusion_insn) -> bool {
+    auto insns = fusion_insn.fused_instructions_computation()->instructions();
+    for (auto insn : insns) {
+      if (insn->shape().layout().format() != DENSE) {
+        continue;
+      }
+      if (!LayoutUtil::IsMonotonicWithDim0Major(insn->shape().layout())) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  for (auto insn : fusion_insns) {
+    if (!is_supported_fusion(*insn)) {
+      VLOG(2) << "Reject unsupported fusion insn " << insn->ToString();
+      continue;
+    } else if (!is_consumer_the_only_user(*insn)) {
+      VLOG(2) << "Reject maybe illegal insn " << insn->ToString();
+      continue;
+    } else if (!is_profitable(*insn)) {
+      VLOG(2) << "Reject may-not-be profitable fusion insn "
+              << insn->ToString();
+      continue;
+    } else if (!all_row_major(*insn)) {
+      VLOG(2) << "Reject non-row-major fusion insn " << insn->ToString();
+      continue;
+    } else {
+      fusion_insns_.push_back(insn);
+    }
+  }
+
+  // Sort according to the number of outputs and instruciton counts.
+  std::sort(fusion_insns_.begin(), fusion_insns_.end(),
```

Comment:
Please motivate the need for sorting in the comment.

Reply:
Will add some comments.